### ISSUE TITLE
should_implement_trait does not consider function safety.

### ISSUE THREAD

**Issue by human-0** [rust-lang/rust-clippy#5413]
Consider this:
```rust
enum Foo(u8) { // Probably struct as well
    _1, _2, ...
}

impl Foo {
    unsafe fn add(self, rhs: u8) -> Foo { ... }
}
```

This triggers `should_implement_trait`, which a false positive, as `Add` requires a normal function, not an unsafe one.

 `cargo clippy -V`:
```
clippy 0.0.212 (4ee12063 2020-02-01)
```

**Comment by flip1995** [comment]
related #4290
